src/librustc_mir/transform/validate.rs

```diff
@@ -83,6 +92,55 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             self.fail(location, format!("encountered jump to invalid basic block {:?}", bb))
         }
     }
+
+    fn check_ty_callable(&self, location: Location, ty: Ty<'tcx>) {
+        if let ty::FnPtr(..) | ty::FnDef(..) = ty.kind {
+            // We have a `FnPtr` or `FnDef` which is trivially safe to call.
```

Comment:
So given the discussion I had with @lcnr, maybe it would be worth adding a comment here that closure calls appear as `FnDef` calls to the `Fn*::call*` method here?

Reply:
Yeah that's a good point - I'll add it in